beacon_node/beacon_chain/src/validator_monitor.rs

```diff
@@ -489,6 +532,152 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+    fn add_validators_missed_blocks(&mut self, state: &BeaconState<T>) {
+        // Define range variables
+        let current_slot = state.slot();
+        let start_slot = current_slot.saturating_sub(T::slots_per_epoch()).as_u64();
+        let end_slot = current_slot.saturating_sub(MISSED_BLOCK_LAG_SLOTS).as_u64();
+
+        // As we are skipping the genesis slot, we can simply pass Hash256::zero() as the shuffling decision block
+        // cf. state.proposer_shuffling_decision_root_at_epoch implementation
+        let shuffling_decision_block = Hash256::zero();
+
+        // List of proposers per epoch from the beacon_proposer_cache
+        let mut proposers_per_epoch: Option<SmallVec<[usize; TYPICAL_SLOTS_PER_EPOCH]>> = None;
+
+        for (prev_slot, slot) in (start_slot..=end_slot).map(Slot::new).tuple_windows() {
+            // Condition for missed_block is defined such as block_root(slot) == block_root(slot - 1)
+            // where the proposer who missed the block is the proposer of the block at block_root(slot)
+            if let (Ok(block_root), Ok(prev_block_root)) =
+                (state.get_block_root(slot), state.get_block_root(prev_slot))
+            {
+                // Found missed block
+                if block_root == prev_block_root {
+                    let slot_epoch = slot.epoch(T::slots_per_epoch());
+                    let prev_slot_epoch = prev_slot.epoch(T::slots_per_epoch());
+
+                    if let Ok(shuffling_decision_block) = state
+                        .proposer_shuffling_decision_root_at_epoch(
+                            slot_epoch,
+                            shuffling_decision_block,
+                        )
+                    {
+                        // Only update the cache if it needs to be initialised or because
+                        // slot is at epoch + 1
+                        if proposers_per_epoch.is_none() || slot_epoch != prev_slot_epoch {
+                            proposers_per_epoch = self.get_proposers_by_epoch_from_cache(
+                                slot_epoch,
+                                shuffling_decision_block,
+                            );
+                        }
+
+                        // Only add missed blocks for the proposer if it's in the list of monitored validators
+                        let slot_in_epoch = slot % T::slots_per_epoch();
+                        if let Some(proposer_index) = proposers_per_epoch
+                            .as_deref()
+                            .and_then(|proposers| proposers.get(slot_in_epoch.as_usize()))
+                        {
+                            let i = *proposer_index as u64;
+                            if let Some(pub_key) = self.indices.get(&i) {
+                                if self.validators.get(pub_key).is_some() {
+                                    // Add to missed blocks
+                                    self.missed_blocks.insert((slot_epoch, i, slot));
```

Comment:
Instead of using the `last_missed_block_validator` concept, I would do something like:

```rust
if self.missed_blocks.insert((slot_epoch, i, slot)).is_none() {
  self.aggregatable_metric(i.to_string().as_str(), |label| {
    metrics::inc_counter_vec(&metrics::VALIDATOR_MONITOR_MISSED_BLOCKS_TOTAL, &[label]);
  });
  // Also, the error log about a missed slot.
}
```

[`HashMap::insert`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.insert) returning `None` indicates that we haven't observed this skipped slot for the given validator before. That does the metrics de-duplication for us.

This does add an edge case though! We're pruning `self.missed_blocks` at the finalised epoch and it's possible that our search back through the last 32 slots might go back into the finalised territory that we've already pruned from `self.missed_blocks` 🙀 We can fix this by pruning everything prior to `finalized_epoch - 1` rather than just `finalized_epoch`. That will give us another epoch of history to work with.

To make things a bit more future-proof with this extra pruning history, I'd suggest defining a `const MISSED_BLOCK_LOOKBACK_EPOCHS = 1`. Then you can use that value for subtracting from `finalized_epoch` *and* for defining `start_slot` on L592.

Perhaps you could add a comment to the definition of `MISSED_BLOCK_LOOKBACK_EPOCHS `  making it clear that setting it to anything higher than `1` is likely going to be problematic because the `beacon_proposer_cache` will be unlikely to contain shuffling for older epochs.